flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/codegen/agg/ImperativeAggCodeGen.scala

```diff
@@ -0,0 +1,502 @@
+import org.apache.flink.table.`type`.{InternalType, RowType, TypeUtils}
+import org.apache.flink.table.codegen.CodeGenUtils._
+import org.apache.flink.table.codegen.GenerateUtils.generateFieldAccess
+import org.apache.flink.table.codegen.agg.AggsHandlerCodeGenerator.{CONTEXT_TERM, CURRENT_KEY, DISTINCT_KEY_TERM, NAMESPACE_TERM, addReusableStateDataViews, createDataViewBackupTerm, createDataViewTerm}
+import org.apache.flink.table.codegen.{CodeGenException, CodeGeneratorContext, ExprCodeGenerator, GeneratedExpression}
+import org.apache.flink.table.dataformat.{GenericRow, UpdatableRow}
+import org.apache.flink.table.dataview.DataViewSpec
+import org.apache.flink.table.expressions.{Expression, ResolvedAggInputReference, ResolvedDistinctKeyReference, RexNodeGenExpressionVisitor}
+import org.apache.flink.table.functions.AggregateFunction
+import org.apache.flink.table.functions.utils.UserDefinedFunctionUtils.{getAggFunctionUDIMethod, getAggUserDefinedInputTypes, getUserDefinedMethod, internalTypesToClasses, signatureToString}
+import org.apache.flink.table.plan.util.AggregateInfo
+import org.apache.flink.table.typeutils.BaseRowTypeInfo
+
+import org.apache.calcite.tools.RelBuilder
+
+import java.lang.reflect.ParameterizedType
+import java.lang.{Iterable => JIterable}
+
+import scala.collection.mutable.ArrayBuffer
+
+/**
+  * It is for code generate aggregation functions that are specified in terms of
+  * accumulate(), retract() and merge() functions. The aggregate accumulator is
+  * embedded inside of a larger shared aggregation buffer.
+  *
+  * @param ctx the code gen context
+  * @param aggInfo  the aggregate information
+  * @param filterExpression filter argument access expression, none if no filter
+  * @param mergedAccOffset the mergedAcc may come from local aggregate,
+  *                        this is the first buffer offset in the row
+  * @param aggBufferOffset  the offset in the buffers of this aggregate
+  * @param aggBufferSize  the total size of aggregate buffers
+  * @param inputTypes   the input field type infos
+  * @param constantExprs  the constant expressions
+  * @param relBuilder  the rel builder to translate expressions to calcite rex nodes
+  * @param hasNamespace  whether the accumulators state has namespace
+  * @param inputFieldCopy    copy input field element if true (only mutable type will be copied)
+  */
+class ImperativeAggCodeGen(
+    ctx: CodeGeneratorContext,
+    aggInfo: AggregateInfo,
+    filterExpression: Option[Expression],
+    mergedAccOffset: Int,
+    aggBufferOffset: Int,
+    aggBufferSize: Int,
+    inputTypes: Seq[InternalType],
+    constantExprs: Seq[GeneratedExpression],
+    relBuilder: RelBuilder,
+    hasNamespace: Boolean,
+    mergedAccOnHeap: Boolean,
+    mergedAccExternalType: TypeInformation[_],
+    inputFieldCopy: Boolean)
+  extends AggCodeGen {
+
+  private val SINGLE_ITERABLE = className[SingleElementIterator[_]]
```

Comment:
The `SingleElementIterator` in flink runtime has a bug:  `iterator` can be called only once after `set` is called. In some aggregate functions (e.g. max_with_retract), the aggregate need to call `accIters` several times.

We have to create another JIRA to fix that or introduce a table's `SingleElementIterator`.

Reply:
ok

Reply:
So the flink-table has its own SingleElementIterator..

Reply:
Yes, `flink-table-planner` has its own `SingleElementIterator` too.

Reply:
First table-blink has own `SingleElementIterator ` too.
created JIRA to fix it: https://issues.apache.org/jira/browse/FLINK-11968